README
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------

                        _____   _   __  __ _____ _____
                       / ____| | | |  \/  |_   _/ ____|
                      | |  __  |_| | \  / | | || |
                      | | |_ |     | |\/| | | || |
                      | |__| |     | |  | |_| || |____
                       \_____|     |_|  |_|_____\_____|

                      GREYC's Magic for Image Computing

                       ( http://gmic.sourceforge.net/ )

                                    _gmic_version

--------------------------------------------------------------------------------

# Summary
#---------

 G'MIC is an open and full-featured framework for image processing, providing
 several different user interfaces to convert/manipulate/filter/visualize
 generic image datasets, from 1d scalar signals to 3d+t sequences of
 multi-spectral volumetric images.

 Technically speaking, what it does is:

 * Define a lightweight but powerful script language (the G'MIC language)
   dedicated to the design of image processing operators and pipelines.

 * Provide several user interfaces embedding the corresponding interpreter:

   - A command-line executable 'gmic', to use the G'MIC framework from a
     shell.
   - A plug-in 'gmic_gimp', to bring G'MIC capabilities to the GIMP
     image retouching software.
   - A web-service 'G'MIC Online', to allow users applying image processing
     algorithms directly in a web browser.
   - A Qt-based interface 'ZArt', for real-time manipulation of webcam images.
   - A thread-safe C++ library 'libgmic', to be linked to third-party applications.

 G'MIC is focused on the design of possibly complex pipelines for converting,
 manipulating, filtering and visualizing generic 1d/2d/3d multi-spectral image
 datasets. This includes of course color images, but also more complex data
 as image sequences or 3d(+t) volumetric float-valued datasets.

 G'MIC is an open framework: the default script language can be extended with
 custom G'MIC-written commands, defining thus new image available filters or
 effects. By the way, it already contains a substantial set of pre-defined image
 processing algorithms and pipelines (more than 1000).

 G'MIC has been designed with portability in mind, and runs on different
 platforms (Windows, Unix, MacOSX). It is distributed under the CeCILL license
 (GPL-compatible). Since 2008, it has been developed in the Image Team of the
 GREYC laboratory, in Caen/France, by permanent researchers working in the field
 of image processing on a daily basis.

# Project Manager and main developer
#-----------------------------------

  David Tschumperlé  ( http://tschumperle.users.greyc.fr/ )

  A complete list of contributors is available on the project web page:

                       http://gmic.sourceforge.net

# Institution
#-------------

 GREYC Image / CNRS UMR 6072 / FRANCE

 Team web page: https://www.greyc.fr/image

# Licenses
#----------

 The C++ source code of G'MIC is distributed under the CECILL v.2.0 license
 (file 'CeCILL.rtf'). This License is a Free-Software license, compatible with
 the GPL (using those files in a distributed closed-source project is then
 forbidden).

 The CeCiLL v.2.0 license ( http://www.cecill.info/index.en.html ) have been
 created under the supervision of the three biggest research institutions on
 computer sciences in France:

   - CNRS  ( http://www.cnrs.fr/ )
   - CEA   ( http://www.cea.fr/ )
   - INRIA ( http://www.inria.fr/ )

# How to install the GIMP plug-in
#---------------------------------

  GIMP (http://www.gimp.org) is an open-source painting and image retouching
  software which can be extended with additional 'plug-ins'.
  G'MIC provides such a plug-in, named 'gmic_gimp'. To install the G'MIC
  plug-in for GIMP, you have to copy the file 'gmic_gimp' (Unix) or
  'gmic_gimp.exe' (Windows) to the plug-in folder of your GIMP installation.

  This folder is usually located at:

  - on Unix: '$HOME/.gimp-2.x/plug-ins/' (for local installation),
             or 'usr/lib/gimp/2.0/plug-ins/' (for global installation).

  - on Windows: 'C:\Program Files\GIMP-2.x\lib\gimp\2.0\plugins\'
                (for global installation, see notes below for other options).

  - on MacOSX: '$HOME/Library/Application\ Support/Gimp/plug-ins/'
               (for installation from GimponOSX Gimp 2.6).

               '$HOME/Library/GIMP/2.8/plug-ins/'
               (for installation from GimponOSX Gimp 2.8+).

  Notes for Windows:

  You may prefer to install the plugin for individual users and avoid having
  to save into the program files folder.  The location differs depending on
  your version of Windows and can be configured in GIMP 2.8, for example:

  - on Windows 7: C:\Users\%USERNAME%\.gimp-2.8\plug-ins

  - on Windows XP: C:\Documents and Settings\%USERNAME%\.gimp-2.8\plug-ins

  See the FAQ for more information.

# More information online
#-------------------------

  - G'MIC wiki pages: http://sourceforge.net/p/gmic/wiki/Home/
  - Help, FAQ: http://sourceforge.net/p/gmic/wiki/FAQ/
  - Forum: http://gimpchat.com/viewforum.php?f=28
  - Reference: http://gmic.sourceforge.net/reference.shtml

--------------------------------------------------------------------------------
--------------------------------------------------------------------------------

debian/changelog
gmic (_gmic_version) testing; urgency=low

  * Current testing release.

 -- David Tschumperlé <David.Tschumperle@ensicaen.fr>  Thu, 30 Dec 2012 11:39:41 +0200

gmic (1.5.0.8+dfsg-1) unstable; urgency=low

  * [b7ddf3a2] Merge commit 'upstream/1.5.0.8+dfsg'
  * [a0dd17c6] Make GMIC build with recent OpenCV versions. (Closes: #652763)
  * [7cb4a04f] Don't use configure, call src/Makefile directly.

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 08 Jan 2012 23:32:01 +0100

gmic (1.5.0.7+dfsg-1) unstable; urgency=low

  * [97e3ecbd] Merge commit 'upstream/1.5.0.7+dfsg'

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 04 Dec 2011 19:55:28 +0100

gmic (1.5.0.5+dfsg-1) unstable; urgency=low

  * [4978288c] Removing Jakub Wilk from Uploaders as requested.
    Thanks for your work Jakub!
  * [b7314e4e] Merge commit 'upstream/1.5.0.5+dfsg'
  * [866550ef] Refactoring debian/rules
  * [fce14609] Bumping Standards-Version, no changes needed.

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 30 Oct 2011 19:06:34 +0100

gmic (1.5.0.0+dfsg-1) unstable; urgency=low

  * [20fc4583] Merge commit 'upstream/1.5.0.0+dfsg'
  * [ec62b62d] Dropping patch for configure fixes, not necessary anymore.
  * [87daddc5] Refreshing patches.
  * [833f51b5] Remove lib*magick++*-dev build-deps, not needed anymore.
  * [7592f26e] Add libopenexr-dev as build-dependency.
  * [5a8d6059] Don't try to install upstream's changelog, not shipped anymore.

 -- Bernd Zeimetz <bzed@debian.org>  Tue, 26 Jul 2011 15:17:46 +0200

gmic (1.4.8.1+dfsg-2) unstable; urgency=low

  * [24dc0e58] Add patch to fix FTBFS on various architectures.
    Thanks to Hector Oron (Closes: #614273)

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 20 Feb 2011 22:52:04 +0100

gmic (1.4.8.1+dfsg-1) unstable; urgency=low

  * [fab6cc4d] Merge commit 'upstream/1.4.8.1+dfsg'

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 16 Feb 2011 22:37:42 +0100

gmic (1.4.8.0+dfsg-1) unstable; urgency=low

  * [33847b44] Merge commit 'upstream/1.4.8.0+dfsg'
  * [6e771817] Refreshing patches.

 -- Bernd Zeimetz <bzed@debian.org>  Thu, 10 Feb 2011 13:13:22 +0100

gmic (1.4.4.2+dfsg-1) experimental; urgency=low

  * [3253ce08] Merge commit 'upstream/1.4.4.2+dfsg'

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 07 Nov 2010 18:47:00 +0100

gmic (1.4.2.0+dfsg-1) experimental; urgency=low

  * [f93deb46] Merge commit 'upstream/1.4.2.0+dfsg'
  * [00290d4f] Refreshing patches.

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 20 Oct 2010 20:57:01 +0200

gmic (1.4.0.0+dfsg-1) experimental; urgency=low

  * [6fd356d5] Merge commit 'upstream/1.4.0.0+dfsg'
  * [331aef59] Adding new build-dependencies.
  * [86b49de0] Refreshing patches.

 -- Bernd Zeimetz <bzed@debian.org>  Fri, 03 Sep 2010 14:49:11 +0200

gmic (1.3.9.0+dfsg-1) unstable; urgency=low

  * [23ccb979] Merge commit 'upstream/1.3.9.0+dfsg'
  * [bef99d90] Install upstream changelog.

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 02 Aug 2010 10:26:51 +0200

gmic (1.3.7.1+dfsg-1) unstable; urgency=low

  * [b573582f] Merge commit 'upstream/1.3.7.1+dfsg'

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 19 Jul 2010 19:11:21 +0200

gmic (1.3.6.0+dfsg-1) unstable; urgency=low

  * [204cd64b] Merge commit 'upstream/1.3.6.0+dfsg'

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 30 Jun 2010 18:42:20 +0200

gmic (1.3.5.7+dfsg-1) unstable; urgency=low

  * [f733c366] Merge commit 'upstream/1.3.5.7+dfsg'
  * [ee1cba19] Refreshing patches.

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 23 Jun 2010 02:23:16 +0200

gmic (1.3.5.4+dfsg-2) unstable; urgency=low

  * [e14a94e2] Rise required amount of RAM to avoid FTBFS on s390.
    (*sigh* - these machines should have neough RAM and swap....)

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 07 Jun 2010 19:37:02 +0200

gmic (1.3.5.4+dfsg-1) unstable; urgency=low

  * [821186d5] Fix dversionmangle in watch file.
  * [8fab3d8a] Merge commit 'upstream/1.3.5.4+dfsg'
    - This icludes a fix for the FTBFS on various archicetures.
      (Closes: #584304)
  * [7550bc35] Refresh patches.
  * [d682b862] Don't gunzip manpages and create directories before make
    install.
  * [6dccd71b] Tidy debian/rules' clean target.
  * [dba56977] Install bash completion

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 07 Jun 2010 13:53:22 +0200

gmic (1.3.5.1+dfsg2-1) unstable; urgency=low

  * Upstream released a bugfix-release for 1.3.5.1 - unfortunately with the
    same version number.
  * [0e506788] Merge commit 'upstream/1.3.5.1+dfsg2'

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 26 May 2010 16:48:40 +0200

gmic (1.3.5.1+dfsg-1) unstable; urgency=low

  * [97366020] Use -O0 in case the system we build on has less than 1GB
    RAM.
  * [6aca1038] Merge commit 'upstream/1.3.5.1+dfsg'
  * [fe6b2057] Fix awk call to evaluate the available memory.
  * [6806172f] Updating patches.

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 26 May 2010 12:13:50 +0200

gmic (1.3.5.0+dfsg-1) unstable; urgency=low

  * [4efc6ba8] Add Enhances/Depends on gimp.
  * [1cec005d] Merge commit 'upstream/1.3.5.0+dfsg'
  * [4a1816e8] Don't use pristine-tar on git-import-orig. We need to
    recreate the tarball after removing non-free material.

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 16 May 2010 22:48:10 +0200

gmic (1.3.5.0~beta1+dfsg-1) experimental; urgency=low

  * Beta version of the upcoming upstream version.
  * [55bc031b] Merge commit 'upstream/1.3.5.0.beta1+dfsg'
  * [35c39744] Bumping Standards-Version to 3.8.4, no changes needed.
  * [1e7913f6] debian/control: Adding VCS Information.

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 28 Apr 2010 19:49:35 +0200

gmic (1.3.4.1+dfsg-4) unstable; urgency=low

  * [973d62d8] Small typo making the kfreebsd fix fail. Fixed now.
  * [cd7344ee] Use -O0 on mips to avoid internal errors in g++.

 -- Bernd Zeimetz <bzed@debian.org>  Tue, 20 Apr 2010 22:58:25 +0200

gmic (1.3.4.1+dfsg-3) unstable; urgency=low

  * [045a2f86] Remove strip commands from src/Makefile. dh_strip takes
    care of stripping if necessary.
  * [d3740bdf] Add patch to make gmic build on kfreebsd.

src/gmic.h
/*
 #
 #  File        : gmic.h
 #                ( C++ header file )
 #
 #  Description : GREYC's Magic for Image Computing
 #                ( http://gmic.sourceforge.net )
 #                This file is also a part of the CImg Library project.
 #                ( http://cimg.sourceforge.net )
 #
 #  Note        : Include this file in your C++ source code, if you
 #                want to use the G'MIC interpreter in your own program.
 #
 #  Copyright   : David Tschumperle
 #                ( http://tschumperle.users.greyc.fr/ )
 #
 #  License     : CeCILL v2.0
 #                ( http://www.cecill.info/licences/Licence_CeCILL_V2-en.html )
 #
 #  This software is governed by the CeCILL  license under French law and
 #  abiding by the rules of distribution of free software.  You can  use,
 #  modify and/ or redistribute the software under the terms of the CeCILL
 #  license as circulated by CEA, CNRS and INRIA at the following URL
 #  "http://www.cecill.info".
 #
 #  As a counterpart to the access to the source code and  rights to copy,
 #  modify and redistribute granted by the license, users are provided only
 #  with a limited warranty  and the software's author,  the holder of the
 #  economic rights,  and the successive licensors  have only  limited
 #  liability.
 #
 #  In this respect, the user's attention is drawn to the risks associated
 #  with loading,  using,  modifying and/or developing or reproducing the
 #  software by the user in light of its specific status of free software,
 #  that may mean  that it is complicated to manipulate,  and  that  also
 #  therefore means  that it is reserved for developers  and  experienced
 #  professionals having in-depth computer knowledge. Users are therefore
 #  encouraged to load and test the software's suitability as regards their
 #  requirements in conditions enabling the security of their systems and/or
 #  data to be ensured and,  more generally, to use and operate it in the
 #  same conditions as regards security.
 #
 #  The fact that you are presently reading this means that you have had
 #  knowledge of the CeCILL license and that you accept its terms.
 #
*/
#include <locale>
#ifndef gmic_version
#define gmic_version 1571

// Define environment variables.
#ifndef gmic_split_compilation
#define gmic_float
#endif // #ifndef gmic_split_compilation
#ifndef gmic_is_beta
#define gmic_is_beta 0
#endif // #ifndef gmic_is_beta
#ifndef cimg_verbosity
#define cimg_verbosity 1
#endif // #ifndef cimg_verbosity
#if defined(cimg_build)
#define cimg_plugin "examples/gmic.cpp"
#define cimg_include_file "../CImg.h"
#elif defined(gmic_build) // #if defined(cimg_build)
#define cimg_plugin "gmic.cpp"
#define cimg_include_file "./CImg.h"
#endif // #if defined(cimg_build)

#ifdef cimg_include_file
#include cimg_include_file
#if cimg_OS==2
#include <process.h>
#elif cimg_OS==1
#include <cerrno>
#endif // #if cimg_OS==2

// Define some special character codes used for replacement in double quoted strings.
const char _dollar = 23, _lbrace = 24, _rbrace = 25, _comma = 26, _dquote = 28, _arobace = 29,
  _newline = 30;

// Replace special characters in a string.
inline char *gmic_strreplace(char *const str) {
  for (char *s = str ; *s; ++s) {
    const char c = *s;
    if (c<' ')
      *s = c==_dollar?'$':c==_lbrace?'{':c==_rbrace?'}':c==_comma?',':
        c==_dquote?'\"':c==_arobace?'@':c;
  }
  return str;
}

#else // #ifdef cimg_include_file
#include <cstdio>
#include <cstring>

#ifndef cimg_version
namespace cimg_library {

  // Define the G'MIC image structure.
  //----------------------------------
  template<typename T> struct CImg {
    unsigned int _width;       // Number of image columns (dimension along the X-axis).
    unsigned int _height;      // Number of image lines (dimension along the Y-axis)
    unsigned int _depth;       // Number of image slices (dimension along the Z-axis).
    unsigned int _spectrum;    // Number of image channels (dimension along the C-axis).
    bool _is_shared;           // Tells if the data buffer is shared by another structure.
    T *_data;                  // Pointer to the first pixel value.

    // Destructor.
    ~CImg();
    // Empty constructor.
    CImg():_width(0),_height(0),_depth(0),_spectrum(0),_is_shared(false),_data(0) {}
    // Use to allocate a new image with specified dimension.
    CImg<T>& assign(const unsigned int w, const unsigned int h=1,
                    const unsigned int d=1, const unsigned int s=1);
  };

  // Define the G'MIC image list structure.
  //---------------------------------------
  template<typename T> struct CImgList {
    unsigned int _width;           // Number of images in the list.
    unsigned int _allocated_width; // Allocated items in the list (must be 2^N and >size).
    CImg<T> *_data;                // Pointer to the first image of the list.

    // Destructor.
    ~CImgList();
    // Empty constructor.
    CImgList():_width(0),_allocated_width(0),_data(0) {}
    // Use to allocate a new image list with specified dimension.
    CImgList<T>& assign(const unsigned int n);
  };
}
#endif // #ifndef cimg_version
#endif // #ifdef cimg_include_file
#define gmic_image cimg_library::CImg
#define gmic_list cimg_library::CImgList
#define gmic_display cimg_library::CImgDisplay

// Define the G'MIC exception class.
//----------------------------------
struct gmic_exception {
  gmic_image<char> _command, _message;

  gmic_exception() {}

  gmic_exception(const char *const command, const char *const message) {
    if (command) {
      _command.assign(std::strlen(command)+1,1,1,1);
      std::strcpy(_command._data,command);
    }
    if (message) {
      _message.assign(std::strlen(message)+1,1,1,1);
      std::strcpy(_message._data,message);
    }
  }

  const char *what() const { // Give the error message returned by the G'MIC interpreter.
    return _message._data?_message._data:"";
  }
  const char *command() const {
    return _command._data?_command._data:"";
  }
};

// Define the G'MIC interpreter class.
//------------------------------------
struct gmic {

  // Constructors - Destructors.
  // Use the methods below to create and run the G'MIC interpreter from your C++ source.

  // Destructor.
  ~gmic();

  // Constructors.
  gmic();

  gmic(const char *const commands_line, const char *const custom_commands=0,
       const bool include_default_commands=true, float *const p_progress=0, int *const p_cancel=0);

  template<typename T>
  gmic(const int argc, const char *const *const argv, gmic_list<T>& images,
       gmic_list<char>& images_names,
       const char *const custom_commands=0, const bool include_default_commands=true,
       float *const p_progress=0, int *const p_cancel=0);

  template<typename T>
  gmic(const char *const commands_line, gmic_list<T>& images, gmic_list<char>& images_names,
       const char *const custom_commands=0, const bool include_default_commands=true,
       float *const p_progress=0, int *const p_cancel=0);

  // Method to call parser on an already constructed gmic object.
  template<typename T>
  gmic& parse(const char *const commands_line,
              gmic_list<T> &images, gmic_list<char> &images_names) {
    return parse(commands_line_to_CImgList(commands_line),images,images_names);
  }

  //--------------------------------------------------------------------------------------
  // All functions below should be considered as 'private' and thus, should not be used
  // in your own C++ source code. Use them at your own risk.
  //--------------------------------------------------------------------------------------
  template<typename T>
  void _gmic(const char *const commands_line, gmic_list<T>& images, gmic_list<char>& images_names,
             const char *const custom_commands, const bool include_default_commands,
             float *const p_progress, int *const p_cancel);

  gmic& add_commands(const char *const data_commands,
                     gmic_list<char> (&commands_names)[256],
                     gmic_list<char> (&commands)[256],
                     gmic_list<char> (&commands_has_arguments)[256]);
  gmic& add_commands(std::FILE *const file,
                     gmic_list<char> (&commands_names)[256],
                     gmic_list<char> (&commands)[256],
                     gmic_list<char> (&commands_has_arguments)[256]);
  gmic_image<char> scope2string(const bool is_last_slash=true) const;
  gmic_image<char> scope2string(const gmic_image<unsigned int>& scope_selection,
                                const bool is_last_slash=true) const;

  gmic_image<unsigned int> selection2cimg(const char *const string, const unsigned int indice_max,
                                          const gmic_list<char>& names,
                                          const char *const command, const bool is_selection,
                                          const bool allow_new_name, gmic_image<char>& new_name);

  gmic_image<char> selection2string(const gmic_image<unsigned int>& selection,
                                    const gmic_list<char>& images_names,
                                    const bool display_selection) const;

  gmic_list<char> commands_line_to_CImgList(const char *const commands_line);

  template<typename T>
  gmic_image<char> substitute_item(const char *const source,
                                   gmic_list<T>& images, gmic_list<char>& images_names,
				   unsigned int (&variables_sizes)[256]);

  gmic& print(const char *format, ...);
  template<typename T>
  gmic& print(const gmic_list<T>& list, const char *format, ...);
  template<typename T>
  gmic& print(const gmic_list<T>& list, const gmic_image<unsigned int>& scope_selection,
	      const char *format, ...);

  gmic& warn(const char *format, ...);
  template<typename T>
  gmic& warn(const gmic_list<T>& list, const char *format, ...);
  template<typename T>
  gmic& warn(const gmic_list<T>& list, const gmic_image<unsigned int>& scope_selection,
             const char *format, ...);

  gmic& error(const char *format, ...);
  template<typename T>
  gmic& error(const gmic_list<T>& list, const char *format, ...);
  template<typename T>
  gmic& error(const char *const command, const gmic_list<T>& list, const char *format, ...);
  template<typename T>
  gmic& error(const gmic_list<T>& list, const gmic_image<unsigned int>& scope_selection,
	      const char *format, ...);
  template<typename T>
  gmic& _arg_error(const gmic_list<T>& list, const char *const command,
		   const char *const argument);

  gmic& debug(const char *format, ...);
  template<typename T>
  gmic& debug(const gmic_list<T>& list, const char *format, ...);

  template<typename T>
  gmic& print_images(const gmic_list<T>& images,
                     const gmic_list<char>& images_names,
                     const gmic_image<unsigned int>& selection,
                     const bool is_header=true);
  template<typename T>
  gmic& display_images(const gmic_list<T>& images,
                       const gmic_list<char>& images_names,
                       const gmic_image<unsigned int>& selection,
                       unsigned int *const XYZ);
  template<typename T>
  gmic& display_plots(const gmic_list<T>& images,
                      const gmic_list<char>& images_names,
                      const gmic_image<unsigned int>& selection,
                      const unsigned int plot_type, const unsigned int vertex_type,
                      const double xmin, const double xmax,
                      const double ymin, const double ymax);
  template<typename T>
  gmic& display_objects3d(const gmic_list<T>& images,
                          const gmic_list<char>& images_names,
                          const gmic_image<unsigned int>& selection);
  template<typename T>
  gmic_image<T>& check_image(const gmic_list<T>& list, gmic_image<T>& img);
  template<typename T>
  const gmic_image<T>& check_image(const gmic_list<T>& list, const gmic_image<T>& img);

  template<typename T>
  gmic& remove_images(gmic_list<T>& images, gmic_list<char>& images_names,
                      const gmic_image<unsigned int>& selection,
                      const unsigned int start, const unsigned int end);
  template<typename T>
  gmic& parse(const gmic_list<char>& commands_line,
	      gmic_list<T> &images, gmic_list<char> &images_names) {
    unsigned int variables_sizes[256] = { 0 };
    unsigned int position = 0;
    std::setlocale(LC_NUMERIC,"C");
    scope.assign(1U);
    scope._data[0].assign(2,1,1,1);
    scope._data[0]._data[0] = '.';
    scope._data[0]._data[1] = 0;
    dowhiles.assign(0U);
    repeatdones.assign(0U);
    status.assign(0U);
    is_start = true;
    is_quit = false;
    is_return = false;
    is_default_type = true;
    *progress = -1;
    return _parse(commands_line,position,images,images_names,variables_sizes);
  }

  template<typename T>
  gmic& _parse(const gmic_list<char>& commands_line, unsigned int& position,
               gmic_list<T> &images, gmic_list<char> &images_names,
               unsigned int (&variables_sizes)[256]);
  gmic& _parse_bool(const gmic_list<char>& commands_line, unsigned int& position,
                    gmic_list<bool>& images, gmic_list<char> &images_names,
                    unsigned int (&variables_sizes)[256]);
  gmic& _parse_uchar(const gmic_list<char>& commands_line, unsigned int& position,
                     gmic_list<unsigned char>& images, gmic_list<char> &images_names,
                     unsigned int (&variables_sizes)[256]);
  gmic& _parse_char(const gmic_list<char>& commands_line, unsigned int& position,
                    gmic_list<char>& images, gmic_list<char> &images_names,
                    unsigned int (&variables_sizes)[256]);
  gmic& _parse_ushort(const gmic_list<char>& commands_line, unsigned int& position,
                      gmic_list<unsigned short>& images, gmic_list<char> &images_names,
                      unsigned int (&variables_sizes)[256]);
  gmic& _parse_short(const gmic_list<char>& commands_line, unsigned int& position,
                     gmic_list<short>& images, gmic_list<char> &images_names,
                     unsigned int (&variables_sizes)[256]);
  gmic& _parse_uint(const gmic_list<char>& commands_line, unsigned int& position,
                    gmic_list<unsigned int>& images, gmic_list<char> &images_names,
                    unsigned int (&variables_sizes)[256]);
  gmic& _parse_int(const gmic_list<char>& commands_line, unsigned int& position,
                   gmic_list<int>& images, gmic_list<char> &images_names,
                   unsigned int (&variables_sizes)[256]);
  gmic& _parse_float(const gmic_list<char>& commands_line, unsigned int& position,
                     gmic_list<float>& images, gmic_list<char> &images_names,
                     unsigned int (&variables_sizes)[256]);
  gmic& _parse_double(const gmic_list<char>& commands_line, unsigned int& position,
                      gmic_list<double>& images, gmic_list<char> &images_names,
                      unsigned int (&variables_sizes)[256]);

  // Internal environment variables.
#if cimg_display!=0
  gmic_display instant_window[10];
#endif // #if cimg_display!=0
  gmic_list<char> commands[256], commands_names[256], commands_has_arguments[256],
    variables[256], variables_names[256], scope;
  gmic_list<unsigned int> dowhiles, repeatdones;
  gmic_image<unsigned char> background3d, light3d;
  gmic_image<float> pose3d;
  gmic_image<char> status;
  float focale3d, light3d_x, light3d_y, light3d_z, specular_lightness3d, specular_shininess3d,
    _progress, *progress;
  bool is_released, is_debug, is_start, is_quit, is_return, is_double3d, is_default_type,
    check_elif;
  int verbosity, render3d, renderd3d;
  volatile int _cancel, *cancel;
  unsigned int nb_carriages;
  unsigned long reference_time;

}; // End of the 'gmic' class.

#endif // #ifndef gmic_version

// Local Variables:
// mode: c++
// End:
